grouper/queue_ordered.go

```diff
@@ -0,0 +1,190 @@
+		return g.stop(g.terminationSignal, signals, errTrace)
+	}
+
+	if signal != nil {
+		return g.stop(signal, signals, errTrace)
+	}
+
+	close(ready)
+
+	signal, errTrace = g.waitForSignal(signals, errTrace)
+	return g.stop(signal, signals, errTrace)
+}
+
+func (g *queueOrdered) validate() error {
+	return g.members.Validate()
+}
+
+func (g *queueOrdered) queuedStart(signals <-chan os.Signal) (os.Signal, ErrorTrace) {
+	for _, member := range g.members {
+		p := ifrit.Background(member)
+		cases := make([]reflect.SelectCase, 0, len(g.pool)+3)
+		for i := 0; i < len(g.pool); i++ {
+			cases = append(cases, reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(g.pool[g.members[i].Name].Wait()),
+			})
+		}
+
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(p.Ready()),
+		})
+
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(p.Wait()),
+		})
+
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(signals),
+		})
+
+		chosen, recv, _ := reflect.Select(cases)
+		g.pool[member.Name] = p
+		switch chosen {
+		case len(cases) - 1:
+			// signals
+			return recv.Interface().(os.Signal), nil
+		case len(cases) - 2:
+			// p.Wait
+			var err error
+			if !recv.IsNil() {
+				err = recv.Interface().(error)
+			}
+			return nil, ErrorTrace{
+				ExitEvent{Member: member, Err: err},
+			}
+		case len(cases) - 3:
+			// p.Ready
+		default:
+			// other member has exited
+			var err error = nil
+			if e := recv.Interface(); e != nil {
+				err = e.(error)
+			}
+			return nil, ErrorTrace{
+				ExitEvent{Member: g.members[chosen], Err: err},
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func (g *queueOrdered) waitForSignal(signals <-chan os.Signal, errTrace ErrorTrace) (os.Signal, ErrorTrace) {
+	cases := make([]reflect.SelectCase, 0, len(g.pool)+1)
+	//for i := len(g.pool) - 1; i >= 0; i-- {
+	for i := 0; i < len(g.pool); i++ {
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(g.pool[g.members[i].Name].Wait()),
+		})
+	}
+	cases = append(cases, reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(signals),
+	})
+
+	chosen, recv, _ := reflect.Select(cases)
+	if chosen == len(cases)-1 {
+		return recv.Interface().(os.Signal), errTrace
+	}
+
+	var err error
+	if !recv.IsNil() {
+		err = recv.Interface().(error)
+	}
+
+	errTrace = append(errTrace, ExitEvent{
+		Member: g.members[chosen],
+		Err:    err,
+	})
+
+	return g.terminationSignal, errTrace
+}
+
+func (g *queueOrdered) stop(signal os.Signal, signals <-chan os.Signal, errTrace ErrorTrace) error {
+	errOccurred := false
+	exited := map[string]struct{}{}
+	if len(errTrace) > 0 {
+		for _, exitEvent := range errTrace {
+			exited[exitEvent.Member.Name] = struct{}{}
+			if exitEvent.Err != nil {
+				errOccurred = true
+			}
+		}
+	}
+
+	for i := 0; i < len(g.pool); i++ {
```

Comment:
^ I assume this is the only change. Seems straight forwards enough. :)

Reply:
thanks @tedsuo yup thats the change and we fixed the comment.